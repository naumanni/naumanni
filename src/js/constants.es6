// Mastodon's max message size
export const MASTODON_MAX_CONTENT_SIZE = 500

export const NAUMANNI_VERSION = `${process.env.NAUMANNI_VERSION || 'development'}`

export const REGEX_PGP_FINGERPRINT = /PGP Key Fingerprint: ([0-9a-fA-F]+)/

export const COLUMN_TIMELINE = 'timeline'
export const COLUMN_FRIENDS = 'friends'
export const COLUMN_TALK = 'talk'
export const COLUMN_NOTIFICATIONS = 'notifications'
export const COLUMN_TAG = 'hashtag'

export const TIMELINE_FEDERATION = 'federation'
export const TIMELINE_LOCAL = 'local'
export const TIMELINE_HOME = 'home'

export const TIMELINE_FILTER_BOOST = 'timeline_filter_boost'
export const TIMELINE_FILTER_REPLY = 'timeline_filter_reply'

export const DIALOG_ADD_ACCOUNT = 'addAccount'
export const DIALOG_AUTHORIZE_ACCOUNT = 'authorizeAccount'
export const DIALOG_MEDIA_VIEWER = 'mediaViewer'
export const DIALOG_GLOBAL_PREFERENCES = 'preferences'
export const DIALOG_USER_DETAIL = 'userDetail'
export const DIALOG_WELCOME = 'welcome'

export const SUBJECT_MIXED = '!!__SUBJECT_MIXED__!!'   // Symbolだとjsonizeできないのでダメ

export const VISIBLITY_DIRECT = 'direct'
export const VISIBLITY_PRIVATE = 'private'
export const VISIBLITY_UNLISTED = 'unlisted'
export const VISIBLITY_PUBLIC = 'public'

export const NOTIFICATION_TYPE_MENTION = 'mention'
export const NOTIFICATION_TYPE_REBLOG = 'reblog'
export const NOTIFICATION_TYPE_FAVOURITE = 'favourite'
export const NOTIFICATION_TYPE_FOLLOW = 'follow'

export const KEY_TAB = 9
export const KEY_ENTER = 13  // 0x0D
export const KEY_ESC = 27    // 0x1B
export const KEY_ARROW_UP = 38
export const KEY_ARROW_DOWN = 40

export const ACCT_PATTERN = '[a-zA-Z0-9_]+@[a-zA-Z0-9\.\-]+[a-zA-Z0-9]+'
export const ACCT_REX = new RegExp(ACCT_PATTERN)
export const MESSAGE_TAG_REX = /--NEM\.([0-9a-f]{8})\.(\d+)\/(\d+)--/

export const STREAM_HOME = 'user'
export const STREAM_LOCAL = 'public:local'
export const STREAM_FEDERATION = 'public'
export const STREAM_TAG = 'hashtag'

// MastodonのWebsocketから送られてくるframeのevent types
export const EVENT_UPDATE = 'update'
export const EVENT_NOTIFICATION = 'notification'

// WebsocketManagerのlistnerに送られるevent type
export const WEBSOCKET_EVENT_ERROR = 'error'
export const WEBSOCKET_EVENT_OPEN = 'open'
export const WEBSOCKET_EVENT_MESSAGE = 'message'
export const WEBSOCKET_EVENT_CLOSE = 'close'

// statusをparseしたあとのToken
export const TOKEN_TEXT = 'text'
export const TOKEN_BREAK = 'break'
export const TOKEN_URL = 'url'
export const TOKEN_MENTION = 'mention'
export const TOKEN_HASHTAG = 'hashtag'
export const TOKEN_EMOJI = 'emoji'

// 最下部からこの位置に来たら、AutoPagingを開始する
export const AUTO_PAGING_MARGIN = 300

// Timelineで普段保持するStatus数
export const MAX_STATUSES = 40

// localStorageのKey
export const STORAGE_KEY_LAST_SEND_FROM = 'naumanni::last_send_from'
export const STORAGE_KEY_PREFERENCES = 'naumanni::preferences'
export const STORAGE_KEY_TOOT_VISIBILITY = 'naumanni::toot_visibility'

export const LOCALES = Object.freeze({
  en: 'English',
  ja: '日本語',
})


// naumanniおすすめのInstance
export const NAUMANNI_PREFERRED_INSTANCES = [
  ['mastodon.social'],
  ['mastodon.cloud'],
  ['mastodon.xyz'],
  ['pawoo.net', 'https://pawoo.net/favicon.png'],
  ['mstdn.jp'],
  ['friends.nico'],
]
