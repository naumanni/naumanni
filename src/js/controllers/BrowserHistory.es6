import pathToRegexp from 'path-to-regexp'


/**
 * Historyのうまい使い方がわからず混乱気味
 */
export default class BrowserHistory {
  /**
   * @constructor
   * @param {Context} context almin context
   * @param {bool} useHash
   */
  constructor(context, useHash=false) {
    this.context = context
    this.useHash = useHash
    this.history = useHash
      ? require('history/createHashHistory').default({basename: '/', hashType: 'hashbang'})
      : require('history/createBrowserHistory').default()
    this.history.listen(::this.onChangeLocation)
    this.routes = {}
  }

  // manipulate history
  push(path) {
    this.history.push(path)
  }

  replace(path) {
    this.history.replace(path)
  }

  back() {
    this.history.goBack()
  }

  // routing
  start() {
    // dispatch current location'
    const {pathname, search, hash} = this.history.location
    this.replace(`${pathname}${search}${hash}`)
  }

  route(name, path, callback) {
    const route = {
      path,
      name,
      callback,
    }
    route.re = pathToRegexp(path, route.keys = [], {sensitive: true, strict: true})
    this.routes[name] = route
  }

  /**
   * 現在、与えられたUrlかどうかを判定する
   * @param {string} url 試したいUrl
   * @return {bool}
   */
  is(url) {
    // TODO: めっちゃ雑
    return this.history.location.pathname === url
  }

  makeUrl(name, params, options={}) {
    const route = this.routes[name]
    const compiled = pathToRegexp.compile(route.path)
    const pathname = compiled(params, {pretty: true})

    if(!options.external)
      return pathname

    let href = this.history.createHref({pathname})
    if(!href.startsWith('/'))
      href = '/' + href
    return `${window.location.origin}${href}`
  }

  // private
  onChangeLocation(location, action) {
    this.onChangeState(location, action)
  }

  onChangeState(location, action) {
    // find route
    for(const route of Object.values(this.routes)) {
      const m = route.re.exec(location.pathname)
      if(!m)
        continue

      const {keys} = route
      const params = {}

      for(let i = 1, len = m.length; i < len; ++i) {
        let key = keys[i - 1]
        if(!m[i])
          continue
        let val = decodeURIComponent(m[i].replace(/\+/g, ' '))
        if(val !== undefined || !(hasOwnProperty.call(params, key.name))) {
          params[key.name] = val
        }
      }

      if(route.callback(this, location, params, action) === false)
        break

      return
    }

    console.error('invalid location', location)
    setTimeout(() => this.replace('/'), 1000)
  }
}
