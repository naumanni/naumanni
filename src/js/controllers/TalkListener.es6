import {EventEmitter} from 'events'

import {
  EVENT_UPDATE, EVENT_NOTIFICATION, NOTIFICATION_TYPE_MENTION,
  STREAM_HOME, VISIBLITY_DIRECT, WEBSOCKET_EVENT_MESSAGE,
  TOKEN_MENTION, TOKEN_TEXT,
} from 'src/constants'
import {Status} from 'src/models'
import WebsocketManager from 'src/controllers/WebsocketManager'
import TimelineData from 'src/infra/TimelineData'

import {decryptBlocks} from 'src/controllers/PGP'

const _STATE_INITIAL = 'initial'
const _STATE_LOADING = 'loading'
const _STATE_WATCHING = 'watching'


class EncryptedMessage {
  constructor(checksum, total) {
    checksum = checksum.toLowerCase()
    require('assert')(checksum.match(/^[0-9a-f]{8}$/), 'invalid checksum')
    require('assert')(0 < total < 99, 'invalid totalnumber')

    this.checksum = checksum
    this.total = total
    this.blocks = new Array(this.total)

    this.decrypting = false
    this.decrypted = null

    require('assert')(!this.hasAllBlocks())
  }

  set(index, statusRef) {
    require('assert')(statusRef)
    if(index < 1 || index >= self.total)
      throw new Error('invalid index')
    this.blocks[index - 1] = statusRef
  }

  hasAllBlocks() {
    for(let idx=0; idx < this.total; ++idx) {
      if(!this.blocks[idx])
        return false
    }
    return true
  }

  async decrypt(privateKey) {
    console.log('decrypt', this)
    if(this.decrypting)
      return
    this.decrypting = true
    const blocks = this.blocks.map((statusRef) => statusRef.resolve().content)
      .map((content) => content
        .replace('</p>', '\n\n')
        .replace('<br(\s+\/)?>', '\n')
        .replace(/<\/?[^>]+(>|$)/g, '')
      )

    try {
      const decrypted = await decryptBlocks(blocks, privateKey)
      this.decrypted = decrypted
      return decrypted
    } catch(e) {
      console.error('decription failed', e)
    }
  }

  get isDecrypted() {
    return this.decrypted
  }
}


class EncryptedStatus {
  constructor(encryptedMessage) {
    this.encryptedMessage = encryptedMessage
    this.primary = this.encryptedMessage.blocks[0].expand()
  }

  get account() {
    const {account} = this.primary
    return account
  }

  get createdAt() {
    const {status} = this.primary
    return status.createdAt
  }

  get fetchedAt() {
    const {status} = this.primary
    return status.fetchedAt
  }

  get uri() {
    const {status} = this.primary
    return status.uri
  }

  getIdByHost(...args) {
    const {status} = this.primary
    return status.getIdByHost(...args)
  }

  get parsedContent() {
    if(!this.encryptedMessage.isDecrypted) {
      return [{type: 'text', text: '復号中...'}]
    } else {
      return [{type: 'text', text: this.encryptedMessage.decrypted}]
    }
  }
}


export class TalkBlock {
  constructor(lastStatus, account) {
    this.statuses = [lastStatus]
    this.account = account
    this.contents = null
  }

  push(newStatus) {
    // require('assert')(newStatus.createdAt.isAfter(this.statuses[0].createdAt))
    this.statuses.push(newStatus)
    this.statuses.sort((a, b) => -Status.compareForTimeline(a, b))
  }

  isMatch(newStatus) {
    // 発言者が違えば違うBlock
    if(newStatus.account !== this.account.uri)
      return false

    // 30分離れていたら違うBlock
    const latestAt = this.statuses[0].createdAt
    if(newStatus.createdAt.isAfter(latestAt.add(30, 'm')))
      return false

    return true
  }
}


/**
 * Talkを管理する
 * TalkControllerでいいんでは?
 */
export default class TalkListener extends EventEmitter {
  static EVENT_CHANGE = 'EVENT_CHANGE'

  constructor(memberAccts) {
    for(const memberAcct of memberAccts) {
      require('assert')(typeof memberAcct === 'string')
    }
    require('assert')(memberAccts.length == 1, 'group talk not implemented')

    super()

    this.state = _STATE_INITIAL
    this.token = null
    // 自分以外のMember Object<Acct, Account>
    this.members = memberAccts.reduce((members, acct) => {
      members[acct] = null
      return members
    }, {})
    this.me = null

    // 各Timelineをほじくる際のMaxID。ページングで使う
    this.statusesMaxId = {}

    this.talk = []
    this.statuses = []
    this.encryptedMessages = {}

    this.removeWebsocketListenerHandler = null
  }

  updateToken(token) {
    if(this.state === _STATE_INITIAL) {
      this.token = token
      this.me = token && token.account

      if(this.token && this.me)
        this.start()
    } else {
      require('assert')(this.token)
      require('assert')(this.me)
      if(this.token.isEqual(token) && this.me.isEqual(token ? token.account : null)) {
        // nothing changed
      } else {
        require('assert')(0, 'not implemented')
      }
    }
  }

  close() {
    if(this.removeWebsocketListenerHandler) {
      this.removeWebsocketListenerHandler()
      this.removeWebsocketListenerHandler = null
    }
  }

  onChange(cb) {
    this.on(this.EVENT_CHANGE, cb)
    return this.removeListener.bind(this, this.EVENT_CHANGE, cb)
  }

  isWatching() {
    return this.state === _STATE_WATCHING
  }

  isLoading() {
    return this.state !== _STATE_WATCHING
  }

  // private
  async start() {
    require('assert')(this.state === _STATE_INITIAL)
    this.state = _STATE_LOADING

    // まず宛先Accountを把握
    await this.loadAccounts()

    // みんなの発言を把握すると同時に、websocketのlistenも開始する
    await Promise.all([
      this.listenWebsocket(),
      this.loadStatuses(this.me),
      ...[Object.values(this.members).map((mem) => this.loadStatuses(mem))],
    ])

    // 読み込み終わったのでWatchを開始する
    this.state = _STATE_WATCHING
    this.emitChange()
  }

  async loadAccounts() {
    const {requester} = this.token
    const responses = await Promise.all(
      Object.keys(this.members)
        .map((acct) => requester.searchAccount({q: acct}))
    )

    responses.forEach(({entities, result}) => {
      result
        .map((uri) => entities.accounts[uri])
        .forEach((account) => {
          if(this.members.hasOwnProperty(account.acct))
            this.members[account.acct] = account
        })
    })

    // アカウント取得できていないやつが入ればエラー
    const johndoes = Object.keys(this.members).filter((acct) => this.members[acct] === null)
    if(johndoes.length) {
      new Error(`Unknown accounts ${johndoes}`)
    }

    this.emitChange()
  }

  // メンバーの発言を読み込んでいく
  async loadStatuses(member) {
    const {host, requester} = this.token

    for(let loop = 0; loop < 3; ++loop) {
      const maxId = this.statusesMaxId[member.acct]
      if(maxId === 0)
        break

      const {entities, result} = await requester.listStatuses({
        id: member.getIdByHost(host),
        limit: 40,
        max_id: maxId,
        visibility: VISIBLITY_DIRECT,
      },
        {token: this.token})

      if(!result.length) {
        this.statusesMaxId[member.acct] = 0
        break
      }

      const statusRefs = TimelineData.mergeStatuses(entities, result)
      this.statusesMaxId[member.acct] = statusRefs[statusRefs.length - 1].resolve().getIdByHost(host)
      this.pushStatusesIfMatched(statusRefs)
    }
  }

  async listenWebsocket() {
    this.removeWebsocketListenerHandler = WebsocketManager.listen(
      this.token.instance.makeStreamingAPIUrl(this.token, STREAM_HOME),
      ::this.onNewMessageReceived
    )
  }

  /**
   * このTalk向けのStatusesを選んで、更新する。変更があればemitChangeもする
   * @param {Status[]} statuses
   * @return {bool} 更新したか?
   */
  pushStatusesIfMatched(statuses) {
    let memberUris = new Set([
      this.me.uri,
      ...Object.values(this.members).map((account) => account.uri),
    ])

    statuses = statuses
      .filter((status) => status.resolve().visibility === VISIBLITY_DIRECT)
      .filter((status) => {
        const targets = new Set([status.accountUri, ...status.resolve().mentions.map((m) => m.url)])

        if(targets.size !== memberUris.size)
          return false
        for(const x of targets)
          if(!memberUris.has(x))
            return false
        return true
      })
      .filter((status) => !this.statuses.find((s) => s.uri === status.uri))
    if(!statuses.length)
      return false

    this.statuses = this.statuses.concat(statuses)
    this.statuses.sort((a, b) => -Status.compareForTimeline(a.resolve(), b.resolve()))

    // rebuild talks
    this.decryptStatuses()
    this.rebuildTalk()
    this.emitChange()

    return true
  }

  /**
   */
  decryptStatuses() {
    for(const statusRef of this.statuses) {
      let status = statusRef.resolve()
      const blockInfo = status.messageBlockInfo

      if(blockInfo) {
        let encryptedMessage = this.encryptedMessages[blockInfo.checksum]
        if(!encryptedMessage) {
          encryptedMessage = this.encryptedMessages[blockInfo.checksum] =
            new EncryptedMessage(blockInfo.checksum, blockInfo.total)
        }
        encryptedMessage.set(blockInfo.index, statusRef)

        if(encryptedMessage.hasAllBlocks()) {
          // decrypt status
          encryptedMessage.decrypt(this.token.privateKey)
            .then(() => {
              this.rebuildTalk()
              this.emitChange()
            })
        }
      }
    }
  }

  /**
   */
  rebuildTalk() {
    const talk = []
    let latestTalkBlock = null
    const pushedEncryptedMessages = new Set()

    for(const statusRef of this.statuses) {
      let status = statusRef.resolve()
      const blockInfo = status.messageBlockInfo

      if(blockInfo) {
        if(pushedEncryptedMessages.has(blockInfo.checksum))
          continue

        const encryptedMessage = this.encryptedMessages[blockInfo.checksum]
        require('assert')(encryptedMessage, 'no encrypt message')
        pushedEncryptedMessages.add(blockInfo.checksum)
        status = new EncryptedStatus(encryptedMessage)
      }

      if(latestTalkBlock && latestTalkBlock.isMatch(status)) {
        latestTalkBlock.push(status)
      } else {
        latestTalkBlock = new TalkBlock(status, statusRef.accountRef.resolve())
        talk.push(latestTalkBlock)
      }
    }

    // make contents
    const targetsAccts = new Set([
      this.me.acct,
      ...Object.values(this.members).map((account) => account.acct),
    ])

    for(const talkBlock of talk) {
      talkBlock.contents = talkBlock.statuses.map((status) => {
        if(status instanceof EncryptedStatus) {
          return {
            key: status.uri,
            parsedContent: status.parsedContent,
            createdAt: status.createdAt,
            encrypted: true,
          }
        } else {
          // 冒頭のmentionだけ省く
          let isHead = true
          let parsedContent = status.parsedContent
            .filter((token) => {
              if(isHead) {
                if(token.type === TOKEN_MENTION && targetsAccts.has(token.acct))
                  return false
                else if(token.type === TOKEN_TEXT && !token.text.trim())
                  return false
                else
                  isHead = false
              }
              return true
            })

          return {
            key: status.uri,
            parsedContent: parsedContent,
            createdAt: status.createdAt,
          }
        }
      })
    }

    // replace
    this.talk = talk
  }

  startWatching() {
    require('assert')(this.state === _STATE_LOADING, 'startWatching(): invalid state')
    this.state = _STATE_WATCHING
    this.emitChange()
  }

  emitChange() {
    this.emit(this.EVENT_CHANGE, this)
  }

  onNewMessageReceived({type, payload}) {
    if(type === WEBSOCKET_EVENT_MESSAGE) {
      const {host, acct} = this.token
      let statusRefs

      if(payload.event === EVENT_UPDATE) {
        // 自分の送ったDirect Messageは steam=user に来る
        const {normalizeStatus} = require('src/api/MastodonAPISpec')
        const {entities, result} = normalizeStatus({result: payload.payload}, host, acct)
        statusRefs = TimelineData.mergeStatuses(entities, [result])
      // これいらんかも
      // } else if(payload.event === EVENT_NOTIFICATION) {
      //   // 送られてきたやつは stream=userに notificationが来る(event=updateももちろん来る)
      //   const {normalizeNotification} = require('src/api/MastodonAPISpec')
      //   const {entities, result} = normalizeNotification(payload.payload, host, acct)
      //   const notification = entities.notifications[result]
      //   statusRefs = TimelineData.mergeStatuses(entities, [notification.status])
      }

      if(statusRefs) {
        this.pushStatusesIfMatched(statusRefs)
      }
    }
  }
}
