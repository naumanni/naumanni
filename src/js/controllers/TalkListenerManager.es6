/* @flow */
import {COLUMN_TALK} from 'src/constants'
import {Account, OAuthToken, UIColumn} from 'src/models'
import ChangeEventEmitter from 'src/utils/EventEmitter'
import TalkListener, {TalkBlock} from './TalkListener'


export class TalkModel {
  isLoading: boolean
  me: ?Account
  members: ?{[acct: string]: Account}
  talk: ?TalkBlock[]

  constructor(
    isLoading: boolean = true,
    me: Account = undefined,
    members: {[acct: string]: Account} = {},
    talk: ?TalkBlock[] = []
  ) {
    this.isLoading = isLoading
    this.me = me
    this.members = members
    this.talk = talk
  }

  toProps() {
    return {
      isLoading: this.isLoading,
      me: this.me,
      members: this.members,
      talk: this.talk,
    }
  }
}


/**
 * TalkColumn(s)がsubscribeするTalkListenerをまとめて管理するレイヤ
 */
class TalkListenerManager extends ChangeEventEmitter {
  listeners: Map<string, TalkListener> = new Map()
  listenerRemovers: Map<string, Function> = new Map()

  // public

  onSubscribeListener(token: OAuthToken, column: UIColumn) {
    const {key, type, params: {to}} = column
    require('assert')(type === COLUMN_TALK)
    const listener = new TalkListener([to])
    this.listenerRemovers.set(
      key,
      listener.onChange(this.onChangeTalk.bind(this, key))
    )
    listener.updateToken(token)
    this.listeners.set(key, listener)

    // initial load
    this.emitChange(key)
  }

  onUnsubscribeListener(column: UIColumn) {
    const {key, type} = column
    require('assert')(type === COLUMN_TALK)

    const remover = this.listenerRemovers.get(key)
    remover && remover()

    const listener = this.listeners.get(key)
    listener && listener.close()

    this.listenerRemovers.delete(key)
    this.listeners.delete(key)
  }

  updateTokenIfNeed(token: OAuthToken, column: UIColumn) {
    const {key, type} = column
    require('assert')(type === COLUMN_TALK)
    const listener = this.listeners.get(key)
    listener && listener.updateToken(token)
  }

  // private

  /**
   * @override
   */
  emitChange(columnKey: string) {
    const listener = this.listeners.get(columnKey)

    if(listener != null) {
      const {me, members, talk} = listener
      const isLoading = listener.isLoading()
      const model = new TalkModel(isLoading, me, members, talk)

      this.emit(this.EVENT_CHANGE, columnKey, model)
    }
  }

  // cb

  onChangeTalk(columnKey: string) {
    this.emitChange(columnKey)
  }
}

export default new TalkListenerManager()
