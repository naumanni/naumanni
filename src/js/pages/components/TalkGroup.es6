/* @flow */
import React from 'react'
import {List} from 'immutable'
import classNames from 'classnames'

import {Account, Attachment} from 'src/models'
import {TalkBlock} from 'src/controllers/TalkListener'
import {TalkBubble, UserIconWithHost} from '../parts'


export class TalkGroupModel {
  _me: Account
  _talkGroup: TalkBlock
  _prevTalkGroup: ?TalkBlock
  _nextTalkGroup: ?TalkBlock

  /**
   * @constructor
   * @param{Account} me
   * @param{TalkBlock} talkGroup
   * @param{TalkBlock} prevTalkGroup
   * @param{TalkBlock} nextTalkGroup
   */
  constructor(me: Account, talkGroup: TalkBlock, prevTalkGroup: ?TalkBlock, nextTalkGroup: ?TalkBlock) {
    this._me = me
    this._talkGroup = talkGroup
    this._prevTalkGroup = prevTalkGroup
    this._nextTalkGroup = nextTalkGroup
  }

  get key(): string {
    return `speak-${this._talkGroup.account.acct}-${this._talkGroup.statuses[0].uri}`
  }

  get isMyTalk(): boolean {
    return this._talkGroup.account.isEqual(this._me)
  }

  get showName(): boolean {
    // memberのtalkgroupは、前のTalkGroupが自分であれば名前を表示しない
    return !this.isMyTalk &&
      !(this._prevTalkGroup && this._prevTalkGroup.account.isEqual(this._talkGroup.account))
  }

  get showAvatar(): boolean {
    // memberのtalkgroupは、次のTalkGroupが自分であればアバターを表示しない
    return !this.isMyTalk &&
      !(this._nextTalkGroup && this._nextTalkGroup.account.isEqual(this._talkGroup.account))
  }
}


export default class TalkGroup extends React.PureComponent {
  props: {
    isMyTalk: boolean,
    showAvatar: boolean,
    showName: boolean,
    talkGroup: TalkBlock,
    onClickHashTag: (string, SyntheticEvent) => void,
    onClickMedia: (e: SyntheticEvent, files: List<Attachment>, idx: number) => void,
  }

  render() {
    const {
      isMyTalk, showName, showAvatar, talkGroup,
      onClickHashTag, onClickMedia,
    } = this.props

    return (
      <div
        className={classNames(
          'talk-talkGroup',
          {'is-me': isMyTalk},
          {'is-member': !isMyTalk},
        )}
      >
        {showName && (
          <div className="talk-speakerName">
            {talkGroup.account.display_name || talkGroup.account.acct}
          </div>
        )}
        {showAvatar && (
          <div className="talk-speakerAvatar">
            <UserIconWithHost account={talkGroup.account} />
          </div>
        )}
        <ul className="talk-talkGroupStatuses">
          {talkGroup.contents.map(({key, mediaFiles, parsedContent, createdAt, encrypted, sendStatus}) => (
            <TalkBubble
              key={key}
              createdAt={createdAt.toDate()}
              isEncrypted={encrypted}
              isMyTalk={isMyTalk}
              mediaFiles={mediaFiles}
              parsedContent={parsedContent}
              sendStatus={sendStatus}
              onClickHashTag={onClickHashTag}
              onClickMedia={onClickMedia}
            />
          ))}
        </ul>
      </div>
    )
  }
}
