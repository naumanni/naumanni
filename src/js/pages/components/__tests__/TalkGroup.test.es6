import React from 'react'
import {shallow} from 'enzyme'
import serializer from 'enzyme-to-json/serializer'

import {TOKEN_MENTION, TOKEN_TEXT} from 'src/constants'
import {TalkBlock} from 'src/controllers/TalkListener'
import {
  talkAccountA, talkAccountB,
  talkStatusA1, talkStatusB1, talkStatusB2, talkStatusA2,
} from './fixtures'
import TalkGroup, {TalkGroupModel} from '../TalkGroup'


expect.addSnapshotSerializer(serializer)


const talkBlock1 = new TalkBlock(talkStatusA1, talkAccountA)
const talkBlock2 = new TalkBlock(talkStatusB1, talkAccountB)
talkBlock2.push(talkStatusB2)
const talkBlock3 = new TalkBlock(talkStatusA2, talkAccountA)


describe('TalkGroupModel', () => {

  describe('own talk', () => {
    const me = talkAccountA
    const model = new TalkGroupModel(me, talkBlock1, undefined, talkBlock2)

      it('`isMyTalk` will be true', () =>  expect(model.isMyTalk).toBeTruthy() )
      it('`showName` will be false', () =>  expect(model.showName).toBeFalsy() )
      it('`showAvatar` will be false', () =>  expect(model.showAvatar).toBeFalsy() )
  })

  describe('member\'s talk', () => {
    const me = talkAccountB
    let model

    it('`isMyTalk` will be false', () => {
      expect(model.isMyTalk).toBeFalsy()
    })

    describe('as new TalkBlock', () => {
      model = new TalkGroupModel(me, talkBlock2, talkBlock1, talkBlock3)
      it('`showName` will be true', () =>  expect(model.showName).toBeTruthy() )
    })

    describe('as last TalkBlock', () => {
      model = new TalkGroupModel(me, talkBlock3, talkBlock2, undefined)
      it('`showAvatar` will be true', () =>  expect(model.showAvatar).toBeTruthy() )
    })
  })
})


describe('TalkGroup', () => {
  const talk = [talkBlock1, talkBlock2, talkBlock3]
  const createShallowWrapper = (me, talkGroup, prevTalkGroup, nextTalkGroup) => {
    const targetsAccts = new Set([talkAccountA, talkAccountB].map(({acct}) => acct))
    if(!talkGroup.contents) {
      talkGroup.contents = talkGroup.statuses.map((status) => {
        // 冒頭のmentionだけ省く
        let isHead = true
        let parsedContent = status.parsedContent
          .filter((token) => {
            if(isHead) {
              if(token.type === TOKEN_MENTION && targetsAccts.has(token.acct))
                return false
              else if(token.type === TOKEN_TEXT && !token.text.trim())
                return false
              else
                isHead = false
            }
            return true
          })

        return {
          key: status.uri,
          parsedContent: parsedContent,
          createdAt: status.createdAt,
        }
      })
    }
    const model = new TalkGroupModel(me, talkGroup, prevTalkGroup, nextTalkGroup)
    const {isMyTalk, showName, showAvatar} = model
    const props = {
      isMyTalk,
      showName,
      showAvatar,
      talkGroup,
      onClickHashTag: jest.fn(),
    }
    return shallow(<TalkGroup {...props} />)
  }

  describe('own talk', () => {
    const me = talkAccountA
    const wrapper = createShallowWrapper(me, talkBlock1, undefined, talkBlock2)
    it('match snapshot', () => expect(wrapper).toMatchSnapshot())
  })

  describe('member\'s talk', () => {
    const me = talkAccountB
    describe('member\'s talk as new TalkBlock', () => {
      const wrapper = createShallowWrapper(me, talkBlock2, talkBlock1, talkBlock3)
      it('match snapshot', () => expect(wrapper).toMatchSnapshot())
    })
    describe('member\'s talk as last TalkBlock', () => {
      const wrapper = createShallowWrapper(me, talkBlock3, talkBlock2, undefined)
      it('match snapshot', () => expect(wrapper).toMatchSnapshot())
    })
  })
})
