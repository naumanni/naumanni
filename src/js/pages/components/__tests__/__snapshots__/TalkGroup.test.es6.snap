// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TalkGroup member's talk member's talk as last TalkBlock match snapshot 1`] = `
<div
  className="talk-talkGroup is-member"
>
  <div
    className="talk-speakerName"
  >
    glpt@friends.nico
  </div>
  <div
    className="talk-speakerAvatar"
  >
    <UserIconWithHost
      account={
        Object {
          "acct": "glpt@friends.nico",
          "avatar": "/avatars/original/missing.png",
          "avatar_static": "/avatars/original/missing.png",
          "created_at": "2017-05-15T08:06:48.470Z",
          "display_name": "",
          "followers_count": 6,
          "following_count": 3,
          "header": "/headers/original/missing.png",
          "header_static": "/headers/original/missing.png",
          "id_by_host": Object {
            "friends.nico": 111166,
          },
          "locked": false,
          "note": "<p></p>",
          "statuses_count": 111,
          "url": "https://friends.nico/@glpt",
          "username": "glpt",
        }
      }
    />
  </div>
  <ul
    className="talk-talkGroupStatuses"
  >
    <TalkBubble
      createdAt={2017-07-28T09:04:53.204Z}
      onClickHashTag={[Function]}
      parsedContent={
        Array [
          Object {
            "tag": "naumanni",
            "type": "hashtag",
            "url": "https://friends.nico/tags/naumanni",
          },
        ]
      }
    />
  </ul>
</div>
`;

exports[`TalkGroup member's talk member's talk as new TalkBlock match snapshot 1`] = `
<div
  className="talk-talkGroup is-me"
>
  <ul
    className="talk-talkGroupStatuses"
  >
    <TalkBubble
      createdAt={2017-05-24T08:03:14.000Z}
      onClickHashTag={[Function]}
      parsedContent={
        Array [
          Object {
            "text": " hi",
            "type": "text",
          },
        ]
      }
    />
    <TalkBubble
      createdAt={2017-07-20T07:32:14.000Z}
      onClickHashTag={[Function]}
      parsedContent={
        Array [
          Object {
            "text": " hello",
            "type": "text",
          },
        ]
      }
    />
  </ul>
</div>
`;

exports[`TalkGroup own talk match snapshot 1`] = `
<div
  className="talk-talkGroup is-me"
>
  <ul
    className="talk-talkGroupStatuses"
  >
    <TalkBubble
      createdAt={2017-05-24T08:02:12.793Z}
      onClickHashTag={[Function]}
      parsedContent={
        Array [
          Object {
            "text": " hello",
            "type": "text",
          },
        ]
      }
    />
  </ul>
</div>
`;
